### PR TITLE
Fix check connections race

### PR DESCRIPTION
**Pull request by mohitkhullar**
Fix check connections race

### CODE REVIEW

File: db/sqlinterfaces.c [hunk 1]
```diff
@@ -4629,8 +4629,11 @@ int check_active_appsock_connections(struct sqlclntstate *clnt)
         int num_retry = 0;
         int rc = -1;
     retry:
+        /* Don't count this connection until we are done with the check */
+        active_appsock_conns--;
```
Comment (mponomar) [inline db/sqlinterfaces.c:4633]: Don't we need appsock_conn_lk here to modify active_appsock_conns? We can also go over the limit if the dirty check fails for a bunch of connections all at once.

Comment (mohitkhullar) [reply db/sqlinterfaces.c:4636]: yes. we have another check and then retry.

File: db/appsock_handler.c [hunk 2]
```diff
@@ -265,6 +268,33 @@ static void appsock_work(struct thdpool *pool, void *work, void *thddata)
...
+    Pthread_mutex_lock(&appsock_conn_lk);
+    if (active_appsock_conns >= bdb_attr_get(thedb->bdb_attr, BDB_ATTR_MAXAPPSOCKSLIMIT)) {
+        if (close_lru_client(active_appsock_conns, NULL, bdb_attr_get(thedb->bdb_attr, BDB_ATTR_MAXAPPSOCKSLIMIT)) != 0) {
+            num_retry++;
+            if (num_retry < 5) {
+                Pthread_mutex_unlock(&appsock_conn_lk);
+                sleep(1);
+                goto retry; 
+            }
+            w->rejected = 1; /* Reject this query. */
```
Comment (mponomar) [inline db/appsock_handler.c:285]: The only reason for rejected seems to be to force-allow admin connections.  We already know at this point that whether it's admin - we can skip this entire process for admin, and reject non-admin connections right at this point.

File: db/appsock_handler.c [hunk 3]
```diff
@@ -265,6 +268,33 @@ static void appsock_work(struct thdpool *pool, void *work, void *thddata)
...
     appsock_work_args_t *w = (appsock_work_args_t *)work;
     int keepsocket = 0;
+    int curr_appsock_conns = 0;
+    w->rejected = 0;
+    int num_retry = 0;
+
+retry:
+    Pthread_mutex_lock(&appsock_conn_lk);
+    if (active_appsock_conns >= bdb_attr_get(thedb->bdb_attr, BDB_ATTR_MAXAPPSOCKSLIMIT)) {
+        if (close_lru_client(active_appsock_conns, NULL, bdb_attr_get(thedb->bdb_attr, BDB_ATTR_MAXAPPSOCKSLIMIT)) != 0) {
```
Comment (mponomar) [inline db/appsock_handler.c:278]: close_lru_client doesn't actually close - it signals the appsock thread to close.  So even if we found a suitable connection, we still end up allowing more than the maximum number during a spike.

File: db/appsock_handler.c [hunk 4]
```diff
@@ -265,6 +268,33 @@ static void appsock_work(struct thdpool *pool, void *work, void *thddata)
...
+            num_retry++;
+            if (num_retry < 5) {
+                Pthread_mutex_unlock(&appsock_conn_lk);
+                sleep(1);
```
Comment (mponomar) [inline db/appsock_handler.c:282]: Maybe we should just reject and not retry?  The API will try on a different node, which is likely to be faster than waiting here.  The API already has logic to back off and retry.

File: db/sqlinterfaces.c [hunk 5]
```diff
@@ -4621,60 +4621,52 @@ __func__, cnonce);
...
-                // lru_clnt->statement_timedout = 1; already done
-                shutdown(fd, SHUT_RD);
-                logmsg(
-                    LOGMSG_WARN,
-                    "%s: Closing least recently used connection fd %d , total "
-                    "%d \n",
-                    __func__, fd, active_appsock_conns - 1);
-                rc = 0;
-            }
+            shutdown(fd, SHUT_RD);
```
Comment (mponomar) [inline db/sqlinterfaces.c:4645]: We always call this with clnt == NULL, so I don't think we can rely on this not finding our own connection.

Comment (mohitkhullar) [reply db/sqlinterfaces.c:4645]: Used with clnt here.. to make it work with other plugins https://github.com/bloomberg/comdb2/pull/2396/files#diff-4932981bbf2c2bd6bf0267d1a6d39173R4664

File: db/appsock_handler.c (hunk 4 above)
Comment (mohitkhullar) [reply db/appsock_handler.c:282]: We don't want client to keep trying and thus increasing the load, we have to give error back.

File: db/appsock_handler.c (hunk 3 above)
Comment (mohitkhullar) [reply db/appsock_handler.c:278]: Yes, the thread doesn't close immediately, however number of actual open connections decreases

File: db/appsock_handler.c (hunk 2 above)
Comment (mohitkhullar) [reply db/appsock_handler.c:285]: Yes, we have to give error back based on plugin

**Pull request merged**